conditions/CephClusterWarningState.json

```diff
@@ -0,0 +1,17 @@
+{
+  "conditions": [
+    {
+      "type": "alert_is_firing",
+      "alert": {
+        "name": "CephClusterWarningState"
+      }
+    }
+  ],
+  "gathering_functions": {
+    "logs_of_namespace": {
+      "namespace": "openshift-storage",
+      "container": "osd",
+      "tail_lines": 500
```

Comment:
They require so many lines, right? Do we know how many bytes that is?

Reply:
~80kB each file... In my tests, I got 3 files, which resulted in 251kB data. So, yeah, maybe it is too big for the archive, right?

@mulbc can we reduce the amount of captured lines? Is there any specific part of the log that you folks are looking for?

Reply:
If any of these alerts fire, we don't necessarily know why. We don't even know for sure it's due to the OSDs producing any issues.
This makes it hard to predict how many log lines are _enough_.

@oritwas - any ideas for a filter? Can we somehow filter to only capture the error messages?

Reply:
ping @oritwas

Reply:
@rluders - FYI Orit is currently traveling and will probably only have time to respond next week. I'll ping her off-Github in case she has a free minute before then.

Reply:
Yes, it is possible to filter only the error messages.
I am traveling currently and will ask for someone from ODF to take a look.

Reply:
> If any of these alerts fire, we don't necessarily know why.

I think we do. Quoting [`CephClusterWarningState`](https://mid998.github.io/ocs-sop/sop/OSD/CephClusterWarningState.html#_overview),
```
Storage cluster is in warning state for more than 10m.
The rook-ceph-mgr job (Prometheus) has been in a warning state for an unacceptable amount of time.
Check for other alerts that would have triggered prior to this one and troubleshoot those alerts first.
```
Same goes for [`CephClusterErrorState`](https://mid998.github.io/ocs-sop/sop/OSD/CephClusterErrorState.html#_overview).

> Can we somehow filter to only capture the error messages?

I'm not familiar with the specification for the JSON structure that the Insights operator ingests, can you point me there? Maybe we can use some existing functionality, if not, then I think a feature request will need to be raised towards the same operator that allows grepping by regex perhaps.

Reply:
Can we use the alert manager logs, they should have all the alerts triggered

Reply:
+1, Alert Manager logs will have that. However, does the Insights operator have functionality to grep them?

Reply:
Maybe I expressed myself poorly or didn't fully understand the issue of using Alertmanager logs here.

The idea of ​​the implementation here is to detect when `CephClusterWarningState` and `CephClusterErrorState` are fired and then collect the logs from the osd container within the openshift-storage namespace. Thus, every time these alerts are triggered we are collecting 500 lines of logs, which is a considerable number and causes us some problems due to the limitation of bytes that we have in our archive. I think, as in other cases, maybe we could apply some Regex to filter some information from this log and search only for some points that can be considered relevant.

We can only filter by lines that match a Regex rule, or that have a match for the regex rule. It is also possible to add an offset from a line that has a specific regex match.

In this way, it would be possible to capture some relevant information that would help to identify and solve the problem.

So I'm not seeing value in capturing Alertmanager logs, or I'm not understanding the reason behind it.

Reply:
> [..] collect the logs from the osd container [..]

I don't think we log fired alerts in the operator deployments themselves. Which container are you referring to here? Can you post a few logs of the same?

What Orit meant was to grep the logs using regex from the Alertmanager pod instead, as all levels of alerts are logged there.

Reply:
cc @rluders

Reply:
@rexagod its on me... I'll provide some sample data today.

Reply:
I added an sample data from `osd` container log at the jira task: https://issues.redhat.com/browse/CCXDEV-8768

Reply:
cc @rexagod @oritwas

Reply:
My primary concern with this approach is the excessive data consumption here. Suppose alert A and B are fired, in that order, with some time gap, and B collects 500 lines of logs, and A collects 250 lines of logs, so the overlapping 250 lines of logs under A and B are extraneous, and end up taking space.

If getting OSD container logs are not a hard requirement, I would suggest going the Alertmanager path, and only collecting, say 10 past lines for context.

Reply:
OK. So, I guess it changes everything that I've been working on so far. I'll need some time to make the changes and test it. Probably I could have something done by next week.

Reply:
Hi, folks.

Regarding this PR, I do still struggling to understand how the Alertmanager would help here. I may need some help to understand it. But first...

Not sure how familiar are you both with Insights Operator (IO) and Conditional Gathering insight, but I can try to summarize it:

- Basically IO gathers information from Cluster
- IO has a functionality, called Conditional Gathering, that gets alerts from Prometheus and triggers some IO gathering functionality to collect some specific information
  - This PR defines the rule for the conditional gathering to be the trigger
    - When `CephClusterWarningState` and/or `CephClusterErrorState` is fired
    - It triggers the IO gathers to collect logs from specified pod and/or container

This is the very basic summary of it.

Currently, I'm getting [this log](https://issues.redhat.com/secure/attachment/12780068/last-500-lines.log) every time that the alert is triggered, and by the definition of the rule, it gets the last 500 lines of the logs. If both alerts were triggered, it'll still get only one log file per pod + container with 500 lines, it will not get two or more files. So, if I get it right, I'm guessing that 

@rexagod concerns [here](https://github.com/RedHatInsights/insights-operator-gathering-conditions/pull/15#discussion_r1018979936) are not a problem, 'cause it will never happen.
However, 500 lines is still a lot for the IO Archive, 'cause we do have an archive limit (for the whole log file that we generate).

Any questions here? So... as I said at the beginning, I'm struggling to understand how the Alertmanager will help in this situation.

If you folks want, we can find a spot for an mtg to discuss and solve it, if we felt that it would be more efficient.

Also, let me ping @oritwas @mulbc here.

Reply:
> If both alerts were triggered, it'll still get only one log file per pod + container with 500 lines, it will not get two or more files.

It's not about the files, but the extraneous content in that one log file. I'd suggest setting up a call with the folks involved.